File: core/src/main/kotlin/br/com/dillmann/nginxignition/core/nginx/configuration/provider/HostRouteSourceCodeFileProvider.kt
package br.com.dillmann.nginxignition.core.nginx.configuration.provider

import br.com.dillmann.nginxignition.core.host.Host
import br.com.dillmann.nginxignition.core.nginx.configuration.NginxConfigurationFileProvider

internal class HostRouteSourceCodeFileProvider: NginxConfigurationFileProvider {
    override suspend fun provide(basePath: String, hosts: List<Host>): List<NginxConfigurationFileProvider.Output> =
        hosts.filter { it.enabled }.flatMap { buildSourceCodeFiles(it) }

    private fun buildSourceCodeFiles(host: Host): List<NginxConfigurationFileProvider.Output> =
        host.routes
            .filter {
                it.enabled &&
                    it.type == Host.RouteType.SOURCE_CODE &&
                    it.sourceCode?.language == Host.SourceCodeLanguage.JAVASCRIPT
            }
            .map {
                NginxConfigurationFileProvider.Output(
                    name = "host-${host.id}-route-${it.priority}.js",
                    contents = it.sourceCode!!.code,
                )
            }
}

File: core/nginx/cfgfiles/host_route_source_code_file_provider.go
package cfgfiles

import (
	"fmt"

	"dillmann.com.br/nginx-ignition/core/common/coreerror"
	"dillmann.com.br/nginx-ignition/core/host"
	"dillmann.com.br/nginx-ignition/i18n"
)

type hostRouteSourceCodeFileProvider struct{}

func newHostRouteSourceCodeFileProvider() *hostRouteSourceCodeFileProvider {
	return &hostRouteSourceCodeFileProvider{}
}

func (p *hostRouteSourceCodeFileProvider) provide(ctx *providerContext) ([]File, error) {
	outputs := make([]File, 0)

	for _, h := range ctx.hosts {
		files, err := p.buildSourceCodeFiles(ctx, &h)
		if err != nil {
			return nil, err
		}

		outputs = append(outputs, files...)
	}

	return outputs, nil
}

func (p *hostRouteSourceCodeFileProvider) buildSourceCodeFiles(
	ctx *providerContext,
	h *host.Host,
) ([]File, error) {
	outputs := make([]File, 0)

	for _, r := range h.Routes {
		if !r.Enabled || r.Type != host.ExecuteCodeRouteType {
			continue
		}

		if ctx.supportedFeatures.RunCodeType == NoneSupportType {
			return nil, coreerror.New(
				i18n.M(ctx.context, i18n.K.NginxErrorHostRouteCodeNotEnabled),
				false,
			)
		}

		if r.SourceCode.Language == host.JavascriptCodeLanguage {
			outputs = append(outputs, File{
				Name:     fmt.Sprintf("host-%s-route-%d.js", h.ID, r.Priority),
				Contents: r.SourceCode.Contents,
			})
		}
	}

	return outputs, nil
}
